oeps/oep-0008.rst

```diff
@@ -0,0 +1,372 @@
+| Created       | 2016-09-09                                |
++---------------+-------------------------------------------+
+
+
+Abstract
+========
+
+The edX ecosystem has grown from a handful of developers working on a few repos
+to many developers working on many repos. As such, questions arise as to tools
+and frameworks to use in unit testing. This proposal is intended to codify best
+practices and to provide a reference for contributors when making choices.
+
+
+Motivation
+==========
+
+Having common best practices for unit testing tools and frameworks across the
+organization's repos will increase the productivity of developers by reducing
+the ramp-up cost when moving among repos. It will also increase the quality of
+the software by decreasing the cognitive load on developers who must call to
+mind the various tips and techniques while using the disparate tools and
+frameworks.
+
+Note that edX already has a few dozen or more existing repos, and they are most
+likely currently not in full compliance with this OEP. This includes the edx-
+platform repo.
+
+Acceptance of this OEP does not imply that we will immediately work to bring all
+repos in full compliance. Rather we will track compliance using the Repository
+Metadata file specified in `OEP-2`_.
+
+Another reason for an exception might be to prototype a proposed new best
+practice. In that case, the openedx.yaml file could be marked as such (using the
+reason field).
+
+.. _OEP-2: https://open-edx-proposals.readthedocs.io/en/latest/oeps/oep-0002.html
+
+
+Specification
+=============
+
+CI (Continuous Integration) Framework
+-------------------------------------
+Our standard is to use `Travis CI`_ for both public and private repos.
+
+We make exceptions for repos which have more complicated CI requirements For
+these we use an internally-managed Jenkins environment.
+
+Examples of special cases for which we use Jenkins rather than Travis:
+
+* The setup and/or execution time for the validation is prohibitively long (e.g.
+  the edx-platform repo)
+* A repo contains secrets that we do not want to share with a third party (e.g.
+  secure repos)
+
+
+Guidelines:
+^^^^^^^^^^^
+
+* Configuration file: ``.travis.yml``
+
+* Use the container-base (sudo-less) CI environment, if possible
+
+* Execute tests in each supported environment
+
+  * For Python projects, execute tests for each supported Python version(s)
+  * For Django, also test in combination with the supported Django version(s)
+
+* Configure with automated deployment of tagged releases to `PyPI`_
+
+
+Task Management
+---------------
+We use Makefiles for task management. The reason for this choice is that
+almost any system no matter the OS should be able to use them without any
+special software installation or configuration.
+
+If the Makefile gets too complex:
+
+* For Python projects, use it as a wrapper to `Paver`_ scripts
+* For JavaScript projectes, use it as a wrapper to `Gulp`_
```

Comment:
In the UXPL we had [some issues](https://openedx.atlassian.net/browse/FEDX-216) stemming from (IMO) the Gulp scripts turning into a dependency rat's nest - synchronous and asynchronous tasks calling each other in a dependency graph that was hard to introspect, through which tasks that supposedly piped output to each other would fail to pass data. (We're now doing this all inside Webpack and have deleted ~60% of the Gulp scripts).

In my projects (personal and at previous jobs) I've used a combination of [npm run-script](https://docs.npmjs.com/cli/run-script) and Webpack to do all the asset management/test running/linting tasks, and I feel like that's covered the bases pretty well.